precepte-core-cats/src/main/scala/package.scala

```diff
@@ -12,62 +12,114 @@ distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
-*/
+ */
 
 package com.mfglabs
 package precepte
 
 import scala.concurrent.Future
-
-import cats.{ Monad, Applicative, Functor, ~> }
+import cats.{Applicative, Functor, Monad, MonadError, ~>}
 import cats.kernel.Semigroup
-import cats.data.{ OptionT, EitherT }
+import cats.data.{EitherT, OptionT}
+import cats.mtl.MonadState
 
 import scala.language.higherKinds
 
 package object corecats extends SubMeta {
 
-  implicit def precepteMonad[Ta, ManagedState, UnmanagedState, F[_]] =
-    new Monad[({ type λ[α] = Precepte[Ta, ManagedState, UnmanagedState, F, α] })#λ] {
-      override def pure[A](a: A): Precepte[Ta, ManagedState, UnmanagedState, F, A] =
-        Return(a)
-      override def map[A, B](m: Precepte[Ta, ManagedState, UnmanagedState, F, A])(f: A => B): Precepte[Ta, ManagedState, UnmanagedState, F, B] =
-        m.map(f)
-      override def flatMap[A, B](m: Precepte[Ta, ManagedState, UnmanagedState, F, A])(f: A => Precepte[Ta, ManagedState, UnmanagedState, F, B]): Precepte[Ta, ManagedState, UnmanagedState, F, B] =
-        m.flatMap(f)
-
-      override def ap[A, B](pab: Precepte[Ta, ManagedState, UnmanagedState, F, A => B])(pa: Precepte[Ta, ManagedState, UnmanagedState, F, A]): Precepte[Ta, ManagedState, UnmanagedState, F, B] = {
-        Apply(pa, pab)
-      }
-      override def tailRecM[A, B](a: A)(f: A => Precepte[Ta, ManagedState, UnmanagedState, F, Either[A,B]]): Precepte[Ta, ManagedState, UnmanagedState, F, B] = {
-        f(a).fastFlatMap {
-          case Left(result) => tailRecM(result)(f)
+  @inline implicit def precepteMonad[T, M, U, F[_]]
+    : (MonadError[Precepte[T, M, U, F, ?], Throwable]
+      with MonadState[Precepte[T, M, U, F, ?], PState[T, M, U]]
+      with cats.Defer[Precepte[T, M, U, F, ?]]) = {
+    class C
+        extends MonadError[Precepte[T, M, U, F, ?], Throwable]
+        with MonadState[Precepte[T, M, U, F, ?], PState[T, M, U]]
+        with cats.Defer[Precepte[T, M, U, F, ?]] {
+
+      final type precepte[A] = Precepte[T, M, U, F, A]
+      final type instrumentStep = SubStepInstrumentation[T, M, U, F]
+      final type state = PState[T, M, U]
+
+      final val monad: Monad[Precepte[T, M, U, F, ?]] = this
+
+      @inline final def inspect[A](
+          f: PState[T, M, U] => A): Precepte[T, M, U, F, A] =
```

Comment:
```suggestion
          f: PState[T, M, U] => A): precepte[A] =
```
?

Shouldn't change anything, mostly for symmetry with the `def`s below.